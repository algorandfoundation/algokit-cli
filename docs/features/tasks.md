# AlgoKit Tasks

AlgoKit Tasks are a collection of handy tasks that can be used to perform various operations on Algorand blockchain.

## Features

- [Wallet Aliasing](./tasks/wallet.md) - Manage your Algorand addresses and accounts effortlessly with the AlgoKit Wallet feature. This feature allows you to create short aliases for your addresses and accounts on AlgoKit CLI.
- [Vanity Address Generation](./tasks/vanity_address.md) - Generate vanity addresses for your Algorand accounts with the AlgoKit Vanity feature. This feature allows you to generate Algorand addresses which contains a specific keyword of your choice.
- [Transfer Assets or Algos](./tasks/transfer.md) - Transfer Algos or Assets from one account to another with the AlgoKit Transfer feature. This feature allows you to transfer Algos or Assets from one account to another on Algorand blockchain.
- [Opt-(in|out) Assets](./tasks/opt.md) - Opt-in or opt-out of Algorand Asset(s). Supports single or multiple assets.
- [Signing transactions](./tasks/sign.md) - Sign goal clerk compatible Algorand transactions.
- [Sending transactions](./tasks/send.md) - Send signed goal clerk compatible Algorand transactions.
- [NFD lookups](./tasks/nfd.md) - Perform a lookup via NFD domain or address, returning the associated address or domain respectively using the AlgoKit CLI.
- [IPFS uploads](./tasks/ipfs.md) - Upload files to IPFS.
- [Asset minting](./tasks/mint.md) - Mint new fungible or non-fungible assets on Algorand.
- [Analyze TEAL code](./tasks/analyze.md) - Analyze TEAL code using [`tealer`](https://github.com/crytic/tealer) integration for common vulnerabilities.
