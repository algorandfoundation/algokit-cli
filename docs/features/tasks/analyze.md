# AlgoKit Task Analyze

The `analyze` task is a command-line utility that analyzes TEAL programs for common vulnerabilities using [Tealer](https://github.com/crytic/tealer) integration. It allows you to detect a range of common vulnerabilities in code written in TEAL. For full list of vulnerability detectors refer to [Tealer documentation](https://github.com/crytic/tealer?tab=readme-ov-file#detectors).

## Usage

```bash
algokit task analyze INPUT_PATHS [OPTIONS]
```

### Arguments

- `INPUT_PATHS`: Paths to the TEAL files or directories containing TEAL files to be analyzed. This argument is required.

### Options

- `-r, --recursive`: Recursively search for all TEAL files within any provided directories.
- `--force`: Force verification without the disclaimer confirmation prompt.
- `--diff`: Exit with a non-zero code if differences are found between current and last reports.
- `-o, --output OUTPUT_PATH`: Directory path where to store the reports of the static analysis.
- `-e, --exclude DETECTORS`: Exclude specific vulnerabilities from the analysis. Supports multiple exclusions in a single run.

## Example

```bash
algokit task analyze ./contracts -r --exclude rekey-to --exclude missing-fee-check
```

This command will recursively analyze all TEAL files in the `contracts` directory and exclude the `missing-fee-check` vulnerability from the analysis.

## Security considerations

This task uses [`tealer`](https://github.com/crytic/tealer), a third-party tool, to suggest improvements for your TEAL programs, but remember to always test your smart contracts code, follow modern software engineering practices and use the [guidelines for smart contract development](https://developer.algorand.org/docs/get-details/dapps/smart-contracts/guidelines/). This should not be used as a substitute for an actual audit.
