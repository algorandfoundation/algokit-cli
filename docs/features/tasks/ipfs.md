# AlgoKit Task IPFS

The AlgoKit IPFS feature allows you to interact with the IPFS [InterPlanetary File System](https://ipfs.tech/) using the [Piñata provider](https://www.pinata.cloud/). This feature supports logging in and out of the Piñata provider, and uploading files to IPFS.

## Usage

Available commands and possible usage as follows:

```bash
$ ~ algokit task ipfs
Usage: algokit task ipfs [OPTIONS]

Upload files to IPFS using Pinata provider.

Options:
  -f, --file PATH Path to the file to upload. [required]
  -n, --name TEXT Human readable name for this upload, for use in file listings.
  -h, --help Show this message and exit.
```

## Options

- `--file, -f PATH`: Specifies the path to the file to upload. This option is required.
- `--name, -n TEXT`: Specifies a human readable name for this upload, for use in file listings.

## Prerequisites

Before you can use this feature, you need to ensure that you have signed up for a Piñata account and have a JWT. You can sign up for a Piñata account by reading [quickstart](https://docs.pinata.cloud/docs/getting-started).

## Login

Please note, you need to login to the Piñata provider before you can upload files. You can do this using the `login` command:

```bash
$ algokit task ipfs login
```

This will prompt you to enter your Piñata JWT. Once you are logged in, you can upload files to IPFS.

## Upload

To upload a file to IPFS, you can use the `ipfs` command as follows:

```bash
$ algokit task ipfs --file {PATH_TO_YOUR_FILE}
```

This will upload the file to IPFS using the Piñata provider and return the CID (Content Identifier) of the uploaded file.

## Logout

If you want to logout from the Piñata provider, you can use the `logout` command:

```bash
$ algokit task ipfs logout
```

This will remove your Piñata JWT from the keyring.

## File Size Limit

Please note, the maximum file size that can be uploaded is 100MB. If you try to upload a file larger than this, you will receive an error.

## Further Reading

For in-depth details, visit the [ipfs section](../../cli/index.md#ipfs) in the AlgoKit CLI reference documentation.
