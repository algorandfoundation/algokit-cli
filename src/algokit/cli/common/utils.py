# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
