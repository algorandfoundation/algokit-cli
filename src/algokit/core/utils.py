import re
import socket


def extract_version_triple(version_str: str) -> str:
    match = re.search(r"\d+\.\d+\.\d+", version_str)
    if not match:
        raise ValueError("Unable to parse version number")
    return match.group()


def is_minimum_version(system_version: str, minimum_version: str) -> bool:
    system_version_as_tuple = tuple(map(int, system_version.split(".")))
    minimum_version_as_tuple = tuple(map(int, minimum_version.split(".")))
    return system_version_as_tuple >= minimum_version_as_tuple


def is_network_available(host: str = "8.8.8.8", port: int = 53, timeout: float = 3.0) -> bool:
    """
    Check if internet is available by trying to establish a socket connection.
    """

    try:
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
        return True
    except OSError:
        return False
