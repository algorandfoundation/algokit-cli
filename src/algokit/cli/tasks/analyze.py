import json
import logging
import re
from pathlib import Path

import click

from algokit.core.tasks.analyze import (
    TEALER_SNAPSHOTS_ROOT,
    generate_report_filename,
    generate_summaries,
    generate_tealer_command,
    has_baseline_diff,
    install_tealer_if_needed,
    load_tealer_report,
    prepare_artifacts_folders,
    run_tealer,
)
from algokit.core.utils import run_with_animation

logger = logging.getLogger(__name__)


def display_analysis_summary(analysis_results: dict) -> None:
    """
    Display the summary of the analysis results.

    Args:
        analysis_results (dict): Dictionary containing analysis results.
    """
    impact_frequency: dict = {}
    for file_path, result_rows in analysis_results.items():
        click.echo(f"\nFile: {file_path}\n")
        for result in result_rows:
            click.echo(
                f"Detector: {result[0]}\n"
                f"Impact: {result[1]}\n"
                f"Details: {result[2]}\n"
                f"Execution Paths (#Lines):\n{result[3]}\n"
            )
            impact_frequency[result[1]] = impact_frequency.get(result[1], 0) + 1
    # print summary by impact label
    click.echo("\nTotal issues:")
    for impact, frequency in impact_frequency.items():
        click.secho(f"{impact}: {frequency}", fg="yellow")


def has_template_vars(path: Path) -> bool:
    """
    Check if the file contains template variables.

    Args:
        path (Path): The file path to check.

    Returns:
        bool: True if template variables are found, False otherwise.
    """
    content = path.read_text()
    return bool(re.search(r"^(?!.*//.*TMPL_).*TMPL_.*", content, flags=re.MULTILINE))


def get_input_files(*, input_paths: tuple[Path], recursive: bool) -> list[Path]:
    """
    Get input files based on the input paths and recursive flag.

    Args:
        input_paths (tuple[Path]): Tuple of input paths.
        recursive (bool): Flag to indicate recursive search.

    Returns:
        list[Path]: List of input files.
    """

    input_files = []
    for input_path in input_paths:
        if input_path.is_dir():
            pattern = "**/*.teal" if recursive else "*.teal"
            input_files.extend(sorted(input_path.glob(pattern)))
        else:
            if recursive:
                click.secho(
                    f"Warning: Ignoring recursive flag for {input_path} as it is not a directory.\n",
                    fg="yellow",
                )
            input_files.append(input_path)
    return sorted(set(input_files))


@click.command(
    name="analyze",
    help=(
        "Analyze TEAL programs for common vulnerabilities using Tealer. "
        "This task uses a third party tool to suggest improvements for your TEAL programs, "
        "but remember to always test your smart contracts code, follow modern software engineering practices "
        "and use the guidelines for smart contract development. "
        "This should not be used as a substitute for an actual audit. "
        "For full list of available detectors, please refer to https://github.com/crytic/tealer?tab=readme-ov-file#detectors"
    ),
)
@click.argument(
    "input_paths",
    nargs=-1,
    type=click.Path(exists=True, dir_okay=True, file_okay=True, path_type=Path),
    required=True,
)
@click.option(
    "-r",
    "--recursive",
    is_flag=True,
    help="Recursively search for all TEAL files within the provided directory.",
)
@click.option(
    "--force",
    is_flag=True,
    help="Force verification without the disclaimer confirmation prompt.",
)
@click.option(
    "--diff",
    "diff_only",
    is_flag=True,
    help=(
        "Exit with a non-zero code if differences are found between current "
        "and last reports. Reports are generated each run, but with this flag "
        "execution fails if the current report doesn't match "
        "the last report. Reports are stored in the "
        ".algokit/static-analysis/snapshots folder by default. Use --output for a "
        "custom path."
    ),
)
@click.option(
    "-o",
    "--output",
    "output_path",
    required=False,
    default=None,
    type=click.Path(dir_okay=True, file_okay=False, resolve_path=True, path_type=Path),
    help=(
        "Directory path where to store the results of the static analysis. "
        "Defaults to .algokit/static-analysis/snapshots."
    ),
)
@click.option(
    "-e",
    "--exclude",
    "detectors_to_exclude",
    multiple=True,
    default=[],
    type=click.STRING,
    help="Exclude specific vulnerabilities from the analysis. Supports multiple exclusions in a single run.",
)
def analyze(  # noqa: PLR0913, C901
    *,
    input_paths: tuple[Path],
    recursive: bool,
    force: bool,
    diff_only: bool,
    output_path: Path | None,
    detectors_to_exclude: list[str],
) -> None:
    """
    Analyze TEAL programs for common vulnerabilities using Tealer.
    """

    # Install tealer if needed
    install_tealer_if_needed()

    detectors_to_exclude = sorted(set(detectors_to_exclude))
    input_files = get_input_files(input_paths=input_paths, recursive=recursive)

    if not force:
        click.confirm(
            click.style(
                "Warning: This task uses `tealer` to suggest improvements for your TEAL programs, "
                "but remember to always test your smart contracts code, follow modern software engineering practices "
                "and use the guidelines for smart contract development. "
                "This should not be used as a substitute for an actual audit. Do you understand?",
                fg="yellow",
            ),
            default=True,
            abort=True,
        )

    reports = {}
    duplicate_files: dict[str, int] = {}
    prepare_artifacts_folders(output_path)
    total_files = len(input_files)
    for index in range(total_files):
        cur_file = input_files[index]
        file = cur_file.resolve()

        if has_template_vars(file):
            click.secho(
                f"Warning: Skipping {file} due to template variables. Substitute them before scanning.",
                err=True,
                fg="yellow",
            )
            continue

        filename = generate_report_filename(file, duplicate_files)

        # If a custom output path is provided, store the report in the specified path
        report_output_root = output_path or TEALER_SNAPSHOTS_ROOT
        report_output_path = report_output_root / filename

        command = generate_tealer_command(cur_file, report_output_path, detectors_to_exclude)
        old_report = load_tealer_report(str(report_output_path)) if report_output_path.exists() and diff_only else None
        if not old_report and diff_only:
            click.secho(
                f"Unable to provide the diff since {file} report is missing. "
                "Please run the task without the --diff flag first.",
                err=True,
                fg="red",
            )
            raise click.exceptions.Exit(1)

        try:
            run_with_animation(run_tealer, f"Analyzing {index + 1} out of {total_files} files", command)

            if diff_only and old_report:
                has_diff = has_baseline_diff(
                    cur_file=cur_file, report_output_path=report_output_path, old_report=old_report
                )
                if has_diff:
                    raise click.exceptions.Exit(1)

            reports[str(report_output_path.absolute())] = json.load(report_output_path.open())
        except Exception as e:
            if diff_only and old_report:
                report_output_path.write_text(json.dumps(old_report.model_dump(by_alias=True), indent=2))

            if isinstance(e, click.exceptions.Exit):
                raise e

            click.secho(
                f"An error occurred while analyzing {cur_file}. "
                "Please make sure the files supplied are valid TEAL code before trying again.",
                err=True,
                fg="red",
            )
            raise click.Abort("Error while running tealer") from e

    summaries = generate_summaries(reports, detectors_to_exclude=detectors_to_exclude)

    if summaries and not diff_only:
        display_analysis_summary(summaries)
        click.echo(f"Finished analyzing {total_files} files.")
        raise click.exceptions.Exit(1)
