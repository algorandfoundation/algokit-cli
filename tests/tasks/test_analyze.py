import re
from collections.abc import Generator
from pathlib import Path
from unittest.mock import MagicMock, patch

import pytest
from algokit.cli.tasks.analyze import has_template_vars
from pytest_mock import MockerFixture

from tests.tasks.conftest import DUMMY_TEAL_FILE_CONTENT
from tests.utils.approvals import verify
from tests.utils.click_invoker import invoke
from tests.utils.proc_mock import ProcMock


def _format_snapshot(output: str, targets: list[str], replacement: str = "dummy") -> str:
    from algokit.core.utils import get_base_python_path

    python_base_path = get_base_python_path()
    if python_base_path is None:
        pytest.fail("Python base detection failed, this should work (even in CI)")

    output = output.replace(python_base_path, "python_base_path")

    for target in targets:
        output = output.replace(target, replacement)

    # If output contains more than one new line trim them to have at most one whitespace in between

    output = re.sub(r"^(pipx:|DEBUG: pipx:).*", "", output, flags=re.MULTILINE)
    return re.sub(r"\n\s*\n", "\n\n", output)


def _normalize_path(path: Path) -> str:
    return str(path.absolute()).replace("\\", r"\\")


@pytest.fixture(autouse=True)
def _disable_animation(mocker: MockerFixture) -> None:
    mocker.patch("algokit.core.utils.animate", return_value=None)


@pytest.fixture(autouse=True)
def cwd(tmp_path_factory: pytest.TempPathFactory) -> Generator[Path, None, None]:
    cwd = tmp_path_factory.mktemp("cwd", numbered=True)

    with patch("algokit.core.tasks.analyze.TEALER_REPORTS_ROOT", return_value=cwd), patch(
        "algokit.core.tasks.analyze.TEALER_SNAPSHOTS_ROOT", return_value=cwd
    ), patch("algokit.core.tasks.analyze.TEALER_DOT_FILES_ROOT", return_value=cwd):
        yield cwd


@pytest.fixture()
def generate_report_filename_mock() -> Generator[MagicMock, None, None]:
    with patch("algokit.cli.tasks.analyze.generate_report_filename", return_value="dummy_report.json") as mock:
        yield mock


@pytest.mark.usefixtures("generate_report_filename_mock")
def test_analyze_single_file(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(f"task analyze {_normalize_path(teal_file)} --output {_normalize_path(cwd)}", input="y\n", cwd=cwd)

    assert result.exit_code == 1
    result.output = _format_snapshot(
        result.output,
        [
            str(cwd),
        ],
    )
    verify(result.output)


def test_analyze_multiple_files(
    cwd: Path,
    generate_report_filename_mock: MagicMock,
) -> None:
    generate_report_filename_mock.side_effect = [f"dummy_{i}.teal" for i in range(5)]
    teal_folder = cwd / "dummy_contracts"
    teal_folder.mkdir()
    for i in range(5):
        teal_file = teal_folder / f"dummy_{i}.teal"
        teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(
        f"task analyze {_normalize_path(teal_folder)} --output {_normalize_path(cwd)}", input="y\n", cwd=cwd
    )

    assert result.exit_code == 1
    for i in range(5):
        result.output = result.output.replace(str(teal_folder / f"dummy_{i}.teal"), f"dummy_contracts/dummy_{i}.teal")
    result.output = _format_snapshot(
        result.output,
        [
            str(cwd),
        ],
    )
    verify(result.output)


def test_analyze_multiple_files_recursive(
    cwd: Path,
    generate_report_filename_mock: MagicMock,
) -> None:
    teal_root_folder = cwd / "dummy_contracts"
    generate_report_filename_mock.side_effect = [teal_root_folder / f"subfolder_{i}/dummy.teal" for i in range(5)]

    for i in range(5):
        teal_folder = teal_root_folder / f"subfolder_{i}"
        teal_folder.mkdir(parents=True)
        teal_file = teal_folder / "dummy.teal"
        teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(
        f"task analyze {_normalize_path(teal_root_folder)} --recursive --output {_normalize_path(cwd)}",
        input="y\n",
        cwd=cwd,
    )

    assert result.exit_code == 1
    for i in range(5):
        result.output = re.sub(r"^File: .*", f"File:  {i}_dummy.teal", result.output, flags=re.MULTILINE)
    result.output = _format_snapshot(
        result.output,
        [str(cwd)],
        "dummy_file.teal",
    )
    verify(result.output)


@pytest.mark.usefixtures("generate_report_filename_mock")
def test_exclude_vulnerabilities(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(
        f"task analyze {_normalize_path(teal_file)} --exclude is-deletable "
        f"--exclude rekey-to --exclude missing-fee-check --output {_normalize_path(cwd)}",
        input="y\n",
        cwd=cwd,
    )

    assert result.exit_code == 0
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


def test_analyze_skipping_tmpl_vars(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(
        DUMMY_TEAL_FILE_CONTENT.replace("pushint 4 // UpdateApplication", "pushint TMPL_VAR // UpdateApplication")
    )
    result = invoke(f"task analyze {_normalize_path(teal_file)}", input="y\n", cwd=cwd)

    assert result.exit_code == 0
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


def test_analyze_commented_tmpl_vars(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(
        DUMMY_TEAL_FILE_CONTENT.replace("pushint 4 // UpdateApplication", "pushint TMPL_VAR // UpdateApplication")
    )
    assert has_template_vars(teal_file)

    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT.replace("pushint 4 // UpdateApplication", "pushint 4 // TMPL_VAR"))
    assert not has_template_vars(teal_file)


def test_analyze_abort_disclaimer(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.touch()
    result = invoke(f"task analyze {_normalize_path(teal_file)} --output {_normalize_path(cwd)}", input="n\n", cwd=cwd)

    assert result.exit_code == 1
    verify(result.output)


def test_analyze_error_in_tealer(
    cwd: Path,
    mocker: MockerFixture,
) -> None:
    mocker.patch("algokit.cli.tasks.analyze.run_tealer", side_effect=Exception("dummy"))
    teal_file = cwd / "dummy.teal"
    teal_file.touch()
    result = invoke(f"task analyze {_normalize_path(teal_file)} --output {_normalize_path(cwd)}", input="y\n", cwd=cwd)

    assert result.exit_code == 1
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


@pytest.mark.usefixtures("generate_report_filename_mock")
def test_analyze_diff_flag(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(f"task analyze {_normalize_path(teal_file)} --output {_normalize_path(cwd)}", input="y\n", cwd=cwd)
    assert result.exit_code == 1

    teal_file.write_text("\n#pragma version 8\nint 1\nreturn\n")
    result = invoke(
        f"task analyze {_normalize_path(teal_file)} --diff --output {_normalize_path(cwd)}", input="y\n", cwd=cwd
    )
    assert result.exit_code == 1
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


@pytest.mark.usefixtures("generate_report_filename_mock")
def test_analyze_diff_flag_missing_old_report(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(
        f"task analyze {_normalize_path(teal_file)} --diff --output {_normalize_path(cwd)}", input="y\n", cwd=cwd
    )
    assert result.exit_code == 1
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


def test_analyze_error_no_pipx(cwd: Path, mocker: MockerFixture, proc_mock: ProcMock) -> None:
    proc_mock.should_fail_on("tealer --version")
    mocker.patch("algokit.core.utils.get_candidate_pipx_commands", return_value=[])

    teal_file = cwd / "dummy.teal"
    teal_file.touch()
    result = invoke(f"task analyze {_normalize_path(teal_file)}", input="y\n", cwd=cwd)

    assert result.exit_code == 1
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)
