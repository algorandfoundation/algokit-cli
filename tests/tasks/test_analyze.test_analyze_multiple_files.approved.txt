DEBUG: Running 'tealer --version' in '{current_working_directory}'
DEBUG: tealer: 0.1.2
Warning: This task uses `tealer` to suggest improvements for your TEAL programs, but remember to always test your smart contracts code, follow modern software engineering practices and use the guidelines for smart contract development. This should not be used as a substitute for an actual audit. Do you understand? [Y/n]: y
DEBUG: Running 'tealer --json {current_working_directory}/dummy_0.teal detect --contracts {current_working_directory}/dummy_contracts/dummy_0.teal' in '{current_working_directory}'
DEBUG: tealer: Reading contract from file: "{current_working_directory}/dummy_contracts/dummy_0.teal"
DEBUG: tealer: json output is written to {current_working_directory}/dummy_0.teal
DEBUG: Running 'tealer --json {current_working_directory}/dummy_1.teal detect --contracts {current_working_directory}/dummy_contracts/dummy_1.teal' in '{current_working_directory}'
DEBUG: tealer: Reading contract from file: "{current_working_directory}/dummy_contracts/dummy_1.teal"
DEBUG: tealer: json output is written to {current_working_directory}/dummy_1.teal
DEBUG: Running 'tealer --json {current_working_directory}/dummy_2.teal detect --contracts {current_working_directory}/dummy_contracts/dummy_2.teal' in '{current_working_directory}'
DEBUG: tealer: Reading contract from file: "{current_working_directory}/dummy_contracts/dummy_2.teal"
DEBUG: tealer: json output is written to {current_working_directory}/dummy_2.teal
DEBUG: Running 'tealer --json {current_working_directory}/dummy_3.teal detect --contracts {current_working_directory}/dummy_contracts/dummy_3.teal' in '{current_working_directory}'
DEBUG: tealer: Reading contract from file: "{current_working_directory}/dummy_contracts/dummy_3.teal"
DEBUG: tealer: json output is written to {current_working_directory}/dummy_3.teal
DEBUG: Running 'tealer --json {current_working_directory}/dummy_4.teal detect --contracts {current_working_directory}/dummy_contracts/dummy_4.teal' in '{current_working_directory}'
DEBUG: tealer: Reading contract from file: "{current_working_directory}/dummy_contracts/dummy_4.teal"
DEBUG: tealer: json output is written to {current_working_directory}/dummy_4.teal

File: dummy_0.teal

Detector: is-deletable
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#deletable-application
Execution Paths (#Lines):
2-8->26-30->31-34->35-38->40-45->76-86->46-47

Detector: missing-fee-check
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#missing-fee-field-validation
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

Detector: rekey-to
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#rekeyable-logicsig
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

File: dummy_1.teal

Detector: is-deletable
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#deletable-application
Execution Paths (#Lines):
2-8->26-30->31-34->35-38->40-45->76-86->46-47

Detector: missing-fee-check
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#missing-fee-field-validation
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

Detector: rekey-to
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#rekeyable-logicsig
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

File: dummy_2.teal

Detector: is-deletable
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#deletable-application
Execution Paths (#Lines):
2-8->26-30->31-34->35-38->40-45->76-86->46-47

Detector: missing-fee-check
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#missing-fee-field-validation
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

Detector: rekey-to
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#rekeyable-logicsig
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

File: dummy_3.teal

Detector: is-deletable
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#deletable-application
Execution Paths (#Lines):
2-8->26-30->31-34->35-38->40-45->76-86->46-47

Detector: missing-fee-check
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#missing-fee-field-validation
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

Detector: rekey-to
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#rekeyable-logicsig
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

File: dummy_4.teal

Detector: is-deletable
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#deletable-application
Execution Paths (#Lines):
2-8->26-30->31-34->35-38->40-45->76-86->46-47

Detector: missing-fee-check
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#missing-fee-field-validation
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

Detector: rekey-to
Impact: High
Details: https://github.com/crytic/tealer/wiki/Detector-Documentation#rekeyable-logicsig
Execution Paths (#Lines):
2-8->9-12->14-23->107-114->89-104->115-120->24-25,
2-8->26-30->31-34->35-38->40-45->76-86->46-47,
2-8->26-30->56-62

Total issues:
High: 15
Finished analyzing 5 files.
